safehttp/plugins/cors/cors.go

```diff
@@ -0,0 +1,208 @@
+//   Type response.
+// - Each CORS request must contain the header "X-Cors: 1" or "Sec-Fetch-Mode: cors".
+// All of this is to prevent XSRF.
+type Interceptor struct {
+	// AllowedOrigins determines which origins should be allowed in the
+	// Access-Control-Allow-Origin header.
+	AllowedOrigins map[string]bool
+	allowedHeaders map[string]bool
+	// ExposedHeaders determines which headers should be set in the
+	// Access-Control-Expose-Headers header. If ExposedHeaders is nil, then the
+	// header is not set.
+	ExposedHeaders []string
+	// AllowCredentials determines if Access-Control-Allow-Credentials should be
+	// set to true when requests containing Cookie headers arrive.
+	AllowCredentials bool
+	// MaxAge sets the Access-Control-Max-Age header.
+	// MaxAge=0 results in MaxAge: 5.
+	MaxAge int
+}
+
+// Default creates a CORS Interceptor with default settings.
+// Those defaults are:
+//   - No Exposed Headers
+//   - No Allowed Headers
+//   - AllowCredentials: false
+//   - MaxAge: 5 seconds
+func Default(allowedOrigins ...string) *Interceptor {
+	ao := map[string]bool{}
+	for _, o := range allowedOrigins {
+		ao[o] = true
+	}
+	return &Interceptor{
+		AllowedOrigins: ao,
+		allowedHeaders: map[string]bool{},
+	}
+}
+
+// SetAllowedHeaders sets which headers should be allowed in the Access-Control-Allow-Headers
+// header. The headers are first canonicalized using textproto.CanonicalMIMEHeaderKey.
+func (it *Interceptor) SetAllowedHeaders(headers ...string) {
+	it.allowedHeaders = map[string]bool{}
+	for _, h := range headers {
+		it.allowedHeaders[textproto.CanonicalMIMEHeaderKey(h)] = true
+	}
+}
+
+// Before handles the incoming request and sets headers or responds with an error
+// accordingly.
+func (it *Interceptor) Before(w safehttp.ResponseWriter, r *safehttp.IncomingRequest) safehttp.Result {
+	origin := r.Header.Get("Origin")
+	if !it.AllowedOrigins[origin] {
+		return w.ClientError(safehttp.StatusForbidden)
+	}
+	h := w.Header()
+	allowOrigin, err := h.Claim("Access-Control-Allow-Origin")
+	if err != nil {
+		return w.ServerError(safehttp.StatusInternalServerError)
+	}
+	vary, err := h.Claim("Vary")
+	if err != nil {
+		return w.ServerError(safehttp.StatusInternalServerError)
+	}
+
+	allowCredentials, err := h.Claim("Access-Control-Allow-Credentials")
+	if err != nil {
+		return w.ServerError(safehttp.StatusInternalServerError)
+	}
+	if r.Header.Get("Cookie") != "" && !it.AllowCredentials {
+		return w.ClientError(safehttp.StatusForbidden)
+	}
+
+	// Used to set headers in preflight() or request() if no error occurs.
+	setHeaders := func() {
+		allowOrigin([]string{origin})
+		vary([]string{"Origin"})
+		if r.Header.Get("Cookie") != "" && it.AllowCredentials {
+			// TODO: handle other credentials than cookies:
+			// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Credentials
+			allowCredentials([]string{"true"})
+		}
```

Comment:
Are you sure this is how it works?

Reply:
Yes, I think so. Could you elaborate on what would be wrong?